src/main/ast-to-doc.js

```diff
@@ -38,36 +38,70 @@ function printAstToDoc(ast, options, alignmentSize = 0) {
   }
 
   const cache = new Map();
+  const path = new AstPath(ast);
 
-  function printGenerically(path, args) {
-    const node = path.getValue();
-
-    const shouldCache = node && typeof node === "object" && args === undefined;
-    if (shouldCache && cache.has(node)) {
-      return cache.get(node);
+  let doc = (function printGenerically(nameOrNames, args) {
+    if (process.env.NODE_ENV !== "production" && nameOrNames === path) {
+      throw new Error("Do not pass `path` to `print` function!");
     }
 
-    let doc = callPluginPrintFunction(path, options, printGenerically, args);
-
-    // We let JSXElement print its comments itself because it adds () around
-    // UnionTypeAnnotation has to align the child without the comments
-    if (
-      !printer.willPrintOwnComments ||
-      !printer.willPrintOwnComments(path, options)
-    ) {
-      // printComments will call the plugin print function and check for
-      // comments to print
-      doc = printComments(path, doc, options);
+    // TODO: Remove support for passing `path` in next major version
+    if (nameOrNames === path || typeof nameOrNames === "undefined") {
+      nameOrNames = [];
     }
 
-    if (shouldCache) {
-      cache.set(node, doc);
-    }
+    const names = Array.isArray(nameOrNames) ? nameOrNames : [nameOrNames];
+    const doc = path.call(() => {
```

Comment:
Again, a new function instance is created on every `printGenerically` call. Can it be extracted to an outer scope?

Reply:
https://github.com/prettier/prettier/pull/10557/commits/fe74439ab36459b44b875c6f877fc1f9943c7013

Reply:
That's bad. I remember getting rid of arrow functions led to a performance boost in `doc-utils`. We should be really careful with them in hot paths.

Reply:
Found that PR: https://github.com/prettier/prettier/pull/4848
Getting rid of arrow functions wasn't the critical part of that optimization, but we need to benchmarks different implementations. It might make sense to assign `args` to an outer-scope variable too.

Reply:
I doubt the performance impact, of course it will be faster a little bit if we don't create a new function but it should be tiny difference, V8 can do a good job on optimize this.


Anyway, I added https://github.com/prettier/prettier/pull/10557/commits/e24087bb3332f5bfa5eabb08a53800b054b314ec (without creating a new arrow function).

I'm not sure how to do the benchmark test, can you help me to test?

https://github.com/prettier/prettier/pull/10557/commits/e24087bb3332f5bfa5eabb08a53800b054b314ec (Callback without args)
https://github.com/prettier/prettier/pull/10557/commits/fe74439ab36459b44b875c6f877fc1f9943c7013 (Extract callback)
https://github.com/prettier/prettier/pull/10557/commits/1328bb87f8a51194f7de9ba40ad6222f65e3ddbe (Before extract callback)
https://github.com/prettier/prettier/commit/edce20a822dd87f437d9c55b557e9b945586db9e (Main branch)

Reply:
Here is the benchmark result

```text
# flow_parser.js from `flow-parser@0.145.0`, 2.9MB

e24087bb3332f5bfa5eabb08a53800b054b314ec (Callback without args)

[debug]   "benchmark": "format x 0.06 ops/sec ±5.73% (5 runs sampled)",
[debug]   "hz": 0.06252269581673485,
[debug]   "ms": 15994.19198

fe74439ab36459b44b875c6f877fc1f9943c7013 (Extract callback)

[debug]   "benchmark": "format x 0.06 ops/sec ±3.23% (5 runs sampled)",
[debug]   "hz": 0.06215052078506828,
[debug]   "ms": 16089.969759999998

1328bb87f8a51194f7de9ba40ad6222f65e3ddbe (Before extract callback)

[debug]   "benchmark": "format x 0.05 ops/sec ±16.00% (5 runs sampled)",
[debug]   "hz": 0.05297934041270051,
[debug]   "ms": 18875.282180000002

edce20a822dd87f437d9c55b557e9b945586db9e (Main branch)

[debug]   "benchmark": "format x 0.05 ops/sec ±31.26% (5 runs sampled)",
[debug]   "hz": 0.04564014458775898,
[debug]   "ms": 21910.535320000003

# doc.js from `prettier@2.2.1`, 115KB


e24087bb3332f5bfa5eabb08a53800b054b314ec (Callback without args)

[debug]   "benchmark": "format x 5.17 ops/sec ±27.92% (20 runs sampled)",
[debug]   "hz": 5.169637385866361,
[debug]   "ms": 193.43716500000002

fe74439ab36459b44b875c6f877fc1f9943c7013 (Extract callback)

[debug]   "benchmark": "format x 5.13 ops/sec ±26.42% (20 runs sampled)",
[debug]   "hz": 5.134979746484509,
[debug]   "ms": 194.742735

1328bb87f8a51194f7de9ba40ad6222f65e3ddbe (Before extract callback)

[debug]   "benchmark": "format x 5.24 ops/sec ±28.54% (20 runs sampled)",
[debug]   "hz": 5.241439713419043,
[debug]   "ms": 190.78727499999997

edce20a822dd87f437d9c55b557e9b945586db9e (Main branch)

[debug]   "benchmark": "format x 6.39 ops/sec ±24.34% (23 runs sampled)",
[debug]   "hz": 6.388479671274391,
[debug]   "ms": 156.5317652173913
```

I really can't tell which one is faster... so I run the e24087bb3332f5bfa5eabb08a53800b054b314ec (Callback without args) on flow_parser.js again, result

```
[debug]   "benchmark": "format x 0.05 ops/sec ±48.52% (5 runs sampled)",
[debug]   "hz": 0.05393845863083009,
[debug]   "ms": 18539.64732
```

Reply:
`flow_parser.js` is a strange file. A big part of it consists of huge strings.

Reply:
`main` performed consistently better in my tests, but only because nodes weren't cached (18cb7fb72).
"Before extract callback" with the cache fix had the same performance as `main`.
Looks like this arrow function optimization doesn't make sense. Except perhaps for reducing the depth of the call stack.

Reply:
Thank you very much, in that case I prefer callback with args.